File: packages/kits/CorpusKit/Sources/CorpusKit/SubSpanScoring.swift
// SubSpanScoring.swift
//
// Transient sub-span max-cosine scoring (MISSION_11X_RECALL_GAP_01 Item 1).
//
// Exposes a bounded, on-the-fly scoring pass over a candidate content set:
// each candidate's effectiveDenseText is segmented into token-window sub-spans,
// each sub-span is embedded via the caller's provider, and the max cosine
// similarity against the query embedding is returned per candidate.
//
// DESIGN INVARIANTS:
//   - Zero persistence: sub-span vectors are computed and DISCARDED immediately.
//   - Compute is bounded by candidate set size (~40), not corpus size.
//   - Segmentation is deterministic and cross-port identical: the alphanumeric-
//     run token-window rule mirrors the standalone-passages `PassageProduction`
//     algorithm (same scalar classification, same sliding window arithmetic).
//     The Rust twin (`sub_span_scoring.rs`) uses identical parameters.
//   - Cosine similarity is computed inline (no VectorKit dependency):
//     pure IEEE-754 arithmetic — same result cross-port within a config.
//   - A candidate absent from the source, a provider that returns an empty
//     float vector, or a text with no tokens is scored 0.0 (safe neutral value).
//
// CONFORMANCE CONTRACT:
//   Sub-span range production for any given (text, windowTokens, overlapTokens)
//   must produce bit-identical UTF-8 byte ranges in Swift and Rust on the same
//   input. The cosineSimilarity formula is NOT four-way bit-identical (float
//   arithmetic is reproducible-within-config, not universally bit-exact), but
//   the SEGMENT TEXT passed to the provider is identical, so provider-identical
//   scores result.
//
// See also:
//   - CorpusContentEngine.scoreSubSpans(query:candidateIDs:) — the actor-level
//     surface that wires this module to the engine's source and default provider.
//   - GLK RecallDirector step 5.8 — the caller that scores the candidate pool.
//
// Rust twin: `rust/src/sub_span_scoring.rs`.

import Foundation
import VectorKit

// MARK: - SubSpanScoring

/// Transient sub-span max-cosine scoring capability.
///
/// `SubSpanScoring` is a CorpusKit SDK primitive: any `CorpusContentSource`
/// implementation (standalone `CorpusDocumentStore` or GLK's LocusKit-backed
/// adapter) feeds it identically. GLK's RecallDirector calls it on
/// `CorpusContentEngine.scoreSubSpans`, but SDK consumers may call
/// `SubSpanScoring.score(query:candidateIDs:source:provider:)` directly.
public enum SubSpanScoring {

    // MARK: - Window parameters

    /// Default token window size for sub-span segmentation.
    ///
    /// 32 tokens covers roughly 20–40 words — a typical sentence or short
    /// paragraph. Short texts (< windowTokens tokens) produce a single span
    /// spanning the whole text; long texts get a sliding window.
    public static let defaultWindowTokens = 32

    /// Default token overlap between adjacent windows.
    ///
    /// 8-token overlap ensures context at window boundaries is not lost.
    public static let defaultOverlapTokens = 8

    // MARK: - Primary scoring entry point

    /// Compute sub-span max-cosine scores for a bounded candidate set.
    ///
    /// For each candidate content ID in `candidateIDs`:
    ///   1. Resolves `effectiveDenseText` via `source.records(for:)`.
    ///   2. Segments the text into token-window sub-spans using the
    ///      alphanumeric-run rule (cross-port identical; see `subSpanRanges`).
    ///   3. Embeds each sub-span via `provider.embedFloat`.
    ///   4. Computes cosine similarity against the pre-embedded `query`.
    ///   5. Returns the MAX cosine across all sub-spans, normalized to [0,1]
    ///      using `(cosine + 1) / 2` — the same convention as the dense lane.
    ///
    /// Candidates absent from the source, candidates where `embedFloat` returns
    /// an empty vector, and candidates whose text has no alphanumeric tokens are
    /// not included in the result dictionary (they contribute 0.0 implicitly).
    ///
    /// - Parameters:
    ///   - query: The query text. Embedded once via `provider.embedFloat`.
    ///   - candidateIDs: Bounded content ID set (typically ~40 from the pool).
    ///   - source: The content source used to resolve `effectiveDenseText`.
    ///   - provider: The embedding provider for both query and sub-span vectors.
    ///   - windowTokens: Tokens per sub-span window (default 32).
    ///   - overlapTokens: Token overlap between windows (default 8).
    /// - Returns: `[CorpusContentID: Float]` — max-cosine ∈ [0,1] per candidate.
    ///   Missing keys implicitly score 0.0.
    public static func score(
        query: String,
        candidateIDs: [CorpusContentID],
        source: any CorpusContentSource,
        provider: any EmbeddingProvider,
        windowTokens: Int = defaultWindowTokens,
        overlapTokens: Int = defaultOverlapTokens
    ) async -> [CorpusContentID: Float] {
        guard !query.isEmpty, !candidateIDs.isEmpty else { return [:] }

        // Embed the query once. If the provider has no float lane, return empty.
        let queryVec: [Float]
        do {
            let result = try await provider.embedFloat(query)
            guard !result.isEmpty else { return [:] }
            queryVec = result
        } catch {
            // Provider opted out (embedFloatVocabMiss, or structural opt-out).
            return [:]
        }

        // Batch-resolve content records to get effectiveDenseText.
        // Uses the default N-serial fallback when the source doesn't override;
        // SQL-backed sources (CorpusDocumentStore, GLK adapter) provide a single
        // WHERE…IN query override.
        let records: [CorpusContentID: CorpusContentRecord]
        do {
            records = try await source.records(for: candidateIDs)
        } catch {
            return [:]
        }

        // Score each resolved candidate.
        var out: [CorpusContentID: Float] = [:]
        out.reserveCapacity(records.count)
        for (id, record) in records {
            let text = record.effectiveDenseText
            guard !text.isEmpty else { continue }

            let ranges = subSpanRanges(
                text: text, windowTokens: windowTokens, overlapTokens: overlapTokens)
            guard !ranges.isEmpty else { continue }

            var maxNorm: Float = 0.0
            let utf8 = text.utf8
            for (spanStart, spanLength) in ranges {
                // Extract the sub-span text from UTF-8 byte offsets.
                guard spanStart >= 0, spanStart + spanLength <= utf8.count else { continue }
                let lo = utf8.index(utf8.startIndex, offsetBy: spanStart)
                let hi = utf8.index(lo, offsetBy: spanLength)
                guard let spanText = String(utf8[lo..<hi]) else { continue }

                let spanVec: [Float]
                do {
                    let result = try await provider.embedFloat(spanText)
                    guard !result.isEmpty else { continue }
                    spanVec = result
                } catch {
                    continue
                }

                let cosine = cosineSimilarity(queryVec, spanVec)
                // Normalize cosine ∈ [−1, 1] to [0, 1]: same convention as the
                // dense lane in RecallDirector (`(similarity + 1) / 2`).
                let norm = max(0, min(1, (cosine + 1) / 2))
                if norm > maxNorm { maxNorm = norm }
            }
            if maxNorm > 0 { out[id] = maxNorm }
        }
        return out
    }

    // MARK: - Segmentation (cross-port identical)

    /// Deterministic alphanumeric-run token-window sub-span ranges over UTF-8 text.
    ///
    /// Token boundaries follow the alphanumeric-run rule: a token is a maximal
    /// contiguous run of scalars that are either Unicode-alphabetic OR ASCII
    /// digits (U+0030–U+0039). The rule is identical to `defaultKeywordTokens`
    /// (CorpusDefaultTokenizer) and to `PassageProduction.passageRanges` in the
    /// standalone-passages feature — the same well-tested cross-port boundary.
    ///
    /// Each window spans from the UTF-8 start byte of its first token through
    /// the UTF-8 end byte of its last token. Adjacent windows overlap by
    /// `overlapTokens` tokens. A text shorter than one full window produces a
    /// single range covering all of its tokens.
    ///
    /// **Cross-port contract:** the Rust twin (`sub_span_scoring::sub_span_ranges`)
    /// uses the same scalar predicate, the same window/stride arithmetic, and the
    /// same UTF-8 byte-offset calculation. Given identical (text, windowTokens,
    /// overlapTokens), both ports return the same (start, length) pairs.
    ///
    /// Exposed `internal` so tests can verify the ranges directly. Callers outside
    /// the module use `score(query:candidateIDs:source:provider:)`.
    ///
    /// - Parameters:
    ///   - text: Source text (UTF-8 encoded; Swift String is always valid UTF-8).
    ///   - windowTokens: Token count per window. Clamped to at least 1.
    ///   - overlapTokens: Token overlap between adjacent windows.
    ///     Clamped to [0, windowTokens − 1].
    /// - Returns: UTF-8 byte-offset ranges `(utf8Start: Int, utf8Length: Int)`.
    ///   Empty when `text` contains no alphanumeric tokens.
    static func subSpanRanges(
        text: String,
        windowTokens: Int,
        overlapTokens: Int
    ) -> [(utf8Start: Int, utf8Length: Int)] {
        let window = max(1, windowTokens)
        let overlap = max(0, min(overlapTokens, window - 1))
        let stride = window - overlap

        // Extract alphanumeric-run token byte ranges.
        // Each token = (startByte: Int, endByte: Int) — half-open [start, end).
        var tokenRanges: [(start: Int, end: Int)] = []
        var offset = 0
        var runStart: Int? = nil
        for scalar in text.unicodeScalars {
            let width = scalar.utf8.count
            let isToken = scalar.properties.isAlphabetic
                || (scalar.value >= 0x30 && scalar.value <= 0x39)
            if isToken {
                if runStart == nil { runStart = offset }
            } else if let start = runStart {
                tokenRanges.append((start, offset))
                runStart = nil
            }
            offset += width
        }
        if let start = runStart { tokenRanges.append((start, offset)) }

        guard !tokenRanges.isEmpty else { return [] }

        // Build sliding token windows.
        var out: [(utf8Start: Int, utf8Length: Int)] = []
        var index = 0
        while index < tokenRanges.count {
            let windowEnd = min(index + window, tokenRanges.count)
            let first = tokenRanges[index]
            let last = tokenRanges[windowEnd - 1]
            out.append((utf8Start: first.start, utf8Length: last.end - first.start))
            if windowEnd == tokenRanges.count { break }
            index += stride
        }
        return out
    }

    // MARK: - Cosine similarity (inline, no VectorKit dependency)

    /// L2-normalised cosine similarity between two equal-length float32 vectors.
    ///
    /// Returns a value in [−1, 1]: 1.0 = identical direction, −1.0 = opposite,
    /// 0.0 = orthogonal. Returns 0.0 when either vector is all-zero (no signal).
    ///
    /// **Why inline:** sub-span scoring is the only call site for on-the-fly
    /// cosine over raw `[Float]` arrays. Adding a VectorKit import only for this
    /// one computation would introduce a dependency edge that doesn't exist
    /// anywhere else in CorpusKit core. The pure arithmetic is tiny and correct.
    ///
    /// **Cross-port:** IEEE-754 float32 `+` and `*` are the same in Swift and
    /// Rust on the same hardware. The result is reproducible-within-config, NOT
    /// four-way bit-identical (the DenseMetric determinism note applies here too).
    ///
    /// Exposed `internal` for testing. Not part of the public SDK surface.
    static func cosineSimilarity(_ a: [Float], _ b: [Float]) -> Float {
        guard a.count == b.count, !a.isEmpty else { return 0.0 }
        var dot: Float = 0
        var normA: Float = 0
        var normB: Float = 0
        for i in 0..<a.count {
            dot   += a[i] * b[i]
            normA += a[i] * a[i]
            normB += b[i] * b[i]
        }
        let denom = normA.squareRoot() * normB.squareRoot()
        guard denom > 0 else { return 0.0 }
        return dot / denom
    }
}

File: packages/kits/CorpusKit/rust/src/sub_span_scoring.rs
//! Transient sub-span max-cosine scoring (MISSION_11X_RECALL_GAP_01 Item 1).
//!
//! Rust twin of Swift `SubSpanScoring.swift`. Implements the same bounded,
//! on-the-fly scoring pass over a candidate content set: each candidate's
//! `effective_dense_text` is segmented into token-window sub-spans, each
//! sub-span is embedded via the caller's provider, and the max cosine
//! similarity against the query embedding is returned per candidate.
//!
//! # Design invariants (shared with Swift twin)
//!
//! - **Zero persistence**: sub-span vectors are computed and discarded
//!   immediately. No storage write, no VectorStore interaction.
//! - **Compute bounded**: by the `SubSpanBudget`, not by the candidate pool
//!   or the corpus. The per-record byte cap bounds the windows one record can
//!   produce and the aggregate window budget bounds the `embed_float` calls
//!   one scoring call can make. Candidates are visited in the caller's order,
//!   so the caller decides which candidates the budget reaches first; the
//!   candidates the budget leaves without a window are reported as
//!   `unscored_ids` and keep their stored signals.
//! - **Cross-port determinism**: `sub_span_ranges` uses the same Unicode
//!   alphabetic + ASCII digit (U+0030–U+0039) scalar predicate as Swift's
//!   `subSpanRanges`, the same sliding window arithmetic, and the same UTF-8
//!   byte-offset calculation. Given identical (text, window_tokens,
//!   overlap_tokens), Swift and Rust produce bit-identical ranges. The byte
//!   cap cuts at the same scalar boundary in both ports, and both ports walk
//!   the candidates in the caller's order, so the same budget truncates the
//!   same candidates.
//! - **Cosine inline**: `cosine_similarity` is pure f32 arithmetic with no
//!   BLAS/NEON dependency — reproducible-within-config, not universally
//!   bit-identical across optimization levels, but functionally equivalent
//!   for the scoring comparison.
//! - **Silent degradation**: if the provider returns Err on `embed_float`,
//!   the candidate is skipped and the function returns an empty outcome. No
//!   panic, no error propagation to the caller.
//!
//! # Cross-port contract
//!
//! The segment text fed to the embedding provider for any given candidate
//! is identical in Swift and Rust for the same (text, window, overlap,
//! budget). Provider-identical scores result. The raw cosine values may
//! differ by float rounding, but the ranking is stable across ports.
//!
//! See also:
//! - `CorpusContentEngine::score_sub_spans` — the engine-level surface.
//! - `Corpus::score_sub_spans` — the chunk-based Corpus surface.
//! - GLK RecallDirector step 5.8 — the caller (both ports).

use crate::content::{CorpusContentId, CorpusContentSource};
use std::collections::HashMap;
use synapsekit::EmbeddingProvider;

// ── Public constants ──────────────────────────────────────────────────────────

/// Default token window size for sub-span segmentation. Mirrors Swift
/// `SubSpanScoring.defaultWindowTokens`.
pub const DEFAULT_WINDOW_TOKENS: usize = 32;

/// Default token overlap between adjacent windows. Mirrors Swift
/// `SubSpanScoring.defaultOverlapTokens`.
pub const DEFAULT_OVERLAP_TOKENS: usize = 8;

// ── Budget ────────────────────────────────────────────────────────────────────

/// The work bound of one scoring call. Mirrors Swift `SubSpanBudget`.
///
/// Without a bound the cost of a call is the sum of every candidate's window
/// count, and a candidate's window count grows with its content: a client
/// that can file large records and issue ordinary searches could force
/// hundreds of thousands of synchronous embedding calls per query while the
/// caller holds the estate coordinator lock. The two limits make the cost of
/// a call a constant of the build, not of the estate.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct SubSpanBudget {
    /// Bytes of a record's dense text the segmenter reads. The cut lands on
    /// the last scalar boundary at or below the cap, so a window never
    /// straddles a partial scalar. Mirrors Swift `maxRecordBytes`.
    pub max_record_bytes: usize,
    /// Aggregate sub-span `embed_float` calls one scoring call may make. The
    /// query embedding is not counted. Mirrors Swift `maxWindows`.
    pub max_windows: usize,
}

impl SubSpanBudget {
    /// 16 KiB per record: covers about nine in ten drawers of a conversation
    /// estate whole (the LME-S aggregate's p90 body is 17 KB) and bounds one
    /// record at about 110 windows under the default 32/8 window.
    pub const DEFAULT_MAX_RECORD_BYTES: usize = 16_384;
    /// 1,024 windows per query: about 1 MB of text through the provider,
    /// roughly a dozen full-cap records, which keeps a query on a small
    /// on-device encoder inside a few seconds under the coordinator lock.
    pub const DEFAULT_MAX_WINDOWS: usize = 1_024;
    /// The default budget. Mirrors Swift `SubSpanBudget.default`.
    pub const DEFAULT: Self = Self {
        max_record_bytes: Self::DEFAULT_MAX_RECORD_BYTES,
        max_windows: Self::DEFAULT_MAX_WINDOWS,
    };
}

impl Default for SubSpanBudget {
    fn default() -> Self {
        Self::DEFAULT
    }
}

/// The result of one scoring call. Mirrors Swift `SubSpanScoringOutcome`.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct SubSpanScoringOutcome {
    /// Max-cosine ∈ [0,1] per candidate the call embedded at least one
    /// window for. Missing keys implicitly score 0.0.
    pub scores: HashMap<CorpusContentId, f32>,
    /// True when the aggregate window budget stopped the call before every
    /// window of every candidate was embedded. The per-record byte cap alone
    /// does not set it: the cap is a constant of the measure, the aggregate
    /// budget is a truncation of the candidate set.
    pub truncated: bool,
    /// The candidates the aggregate budget left without a single embedded
    /// window, in the caller's order. Their dense column stays at its stored
    /// value; the recall explainer names them.
    pub unscored_ids: Vec<CorpusContentId>,
    /// Sub-span `embed_float` calls the call made.
    pub windows_embedded: usize,
}

// ── Primary scoring entry point ───────────────────────────────────────────────

/// Compute sub-span max-cosine scores for a candidate set under a budget.
///
/// For each candidate content ID in `candidate_ids`, in that order:
///   1. Resolves `effective_dense_text` via `source.records_for(ids)`.
///   2. Cuts the text at `budget.max_record_bytes` on a scalar boundary.
///   3. Segments the text into token-window sub-spans using the
///      alphanumeric-run rule (`sub_span_ranges` — cross-port identical).
///   4. Embeds each sub-span via `provider.embed_float` while the aggregate
///      window budget lasts.
///   5. Computes cosine similarity against the pre-embedded `query`.
///   6. Returns the MAX cosine across the embedded sub-spans, normalized to
///      [0,1] using `(cosine + 1) / 2` — the same convention as the dense lane.
///
/// Candidates absent from the source, candidates where `embed_float` returns
/// `Err`, and candidates whose text has no alphanumeric tokens are not
/// included in the returned scores (they contribute 0.0 implicitly). A
/// candidate the budget reached only partway is scored over the windows it
/// got; a candidate the budget did not reach at all is listed in
/// `unscored_ids`.
///
/// # Parameters
/// - `query`: The query text. Embedded once via `provider.embed_float`.
/// - `candidate_ids`: Content IDs in priority order: the budget serves the
///   front of the slice first.
/// - `source`: The content source for resolving `effective_dense_text`.
/// - `provider`: The embedding provider for query and sub-span vectors.
/// - `window_tokens`: Tokens per sub-span window.
/// - `overlap_tokens`: Token overlap between windows.
/// - `budget`: The per-record byte cap and the aggregate window budget.
pub fn score(
    query: &str,
    candidate_ids: &[&str],
    source: &dyn CorpusContentSource,
    provider: &dyn EmbeddingProvider,
    window_tokens: usize,
    overlap_tokens: usize,
    budget: SubSpanBudget,
) -> SubSpanScoringOutcome {
    if query.is_empty() || candidate_ids.is_empty() {
        return SubSpanScoringOutcome::default();
    }

    // Embed the query once. If the provider opts out of the float lane, return empty.
    let query_vec = match provider.embed_float(query) {
        Ok(v) if !v.is_empty() => v,
        _ => return SubSpanScoringOutcome::default(),
    };

    // Batch-resolve content records. Falls back to N serial record() calls
    // via the CorpusContentSource default implementation.
    let records = match source.records_for(candidate_ids) {
        Ok(r) => r,
        Err(_) => return SubSpanScoringOutcome::default(),
    };

    let mut outcome = SubSpanScoringOutcome {
        scores: HashMap::with_capacity(records.len()),
        ..SubSpanScoringOutcome::default()
    };
    // The candidate slice, not the record map, drives the walk: the map has
    // no order, and the budget must reach the caller's first candidates first.
    for &id in candidate_ids {
        let Some(record) = records.get(id) else { continue };
        let text = capped_text(record.effective_dense_text(), budget.max_record_bytes);
        if text.is_empty() {
            continue;
        }
        if outcome.windows_embedded >= budget.max_windows {
            // Budget exhausted by an earlier candidate: this one keeps its
            // stored signals. The ranges are not computed; the text is
            // non-empty, which is the condition the caller can act on.
            outcome.truncated = true;
            outcome.unscored_ids.push(id.to_string());
            continue;
        }
        let ranges = sub_span_ranges(text, window_tokens, overlap_tokens);
        if ranges.is_empty() {
            continue;
        }
        let text_bytes = text.as_bytes();
        let mut max_norm: f32 = 0.0;
        let mut embedded_here = 0usize;
        for (span_start, span_length) in &ranges {
            if outcome.windows_embedded >= budget.max_windows {
                outcome.truncated = true;
                break;
            }
            let lo = *span_start;
            let hi = lo + span_length;
            // Byte-range guard (should always hold for well-formed input).
            if hi > text_bytes.len() {
                continue;
            }
            let span_text = match std::str::from_utf8(&text_bytes[lo..hi]) {
                Ok(s) => s,
                Err(_) => continue,
            };
            outcome.windows_embedded += 1;
            embedded_here += 1;
            let span_vec = match provider.embed_float(span_text) {
                Ok(v) if !v.is_empty() => v,
                _ => continue,
            };
            let cosine = cosine_similarity(&query_vec, &span_vec);
            // Normalize cosine ∈ [−1, 1] to [0, 1]: same convention as the
            // dense lane in RecallDirector (`(similarity + 1) / 2`).
            let norm = f32::max(0.0, f32::min(1.0, (cosine + 1.0) / 2.0));
            if norm > max_norm {
                max_norm = norm;
            }
        }
        if embedded_here == 0 {
            outcome.unscored_ids.push(id.to_string());
            continue;
        }
        if max_norm > 0.0 {
            outcome.scores.insert(id.to_string(), max_norm);
        }
    }
    outcome
}

/// The longest prefix of `text` that is at most `max_bytes` long and ends on
/// a scalar boundary. Mirrors Swift `SubSpanScoring.cappedText`: both ports
/// step back from the cap to the nearest boundary, so the same bytes reach
/// the segmenter.
pub fn capped_text(text: &str, max_bytes: usize) -> &str {
    if text.len() <= max_bytes {
        return text;
    }
    let mut cut = max_bytes;
    while cut > 0 && !text.is_char_boundary(cut) {
        cut -= 1;
    }
    &text[..cut]
}

// ── Segmentation (cross-port identical) ──────────────────────────────────────

/// Deterministic alphanumeric-run token-window sub-span ranges over UTF-8 text.
///
/// # Cross-port contract
///
/// The token boundary rule is:
/// - A scalar is in-token iff it is Unicode-alphabetic OR is an ASCII digit
///   (U+0030–U+0039, i.e. `char::is_ascii_digit()`).
/// - The same rule is used by Swift `subSpanRanges`, `defaultKeywordTokens`
///   (CorpusDefaultTokenizer), and `PassageProduction.passageRanges` in the
///   standalone-passages feature.
///
/// Each window spans from the UTF-8 start byte of its first token through
/// the UTF-8 end byte of its last token. Adjacent windows overlap by
/// `overlap_tokens` tokens. A text shorter than one full window produces a
/// single range covering all of its tokens.
///
/// Given identical (text, window_tokens, overlap_tokens), this function and
/// Swift `SubSpanScoring.subSpanRanges` return the same (start, length)
/// pairs.
///
/// # Parameters
/// - `text`: Source text (UTF-8).
/// - `window_tokens`: Token count per window. Clamped to at least 1.
/// - `overlap_tokens`: Token overlap between adjacent windows. Clamped to
///   `[0, window_tokens − 1]`.
///
/// # Returns
/// `Vec<(usize, usize)>` of `(utf8_start, utf8_length)` pairs. Empty when
/// `text` contains no alphanumeric tokens.
pub fn sub_span_ranges(
    text: &str,
    window_tokens: usize,
    overlap_tokens: usize,
) -> Vec<(usize, usize)> {
    let window = window_tokens.max(1);
    let overlap = overlap_tokens.min(window - 1);
    let stride = window - overlap;

    // Collect alphanumeric-run token byte ranges: (start, end) half-open.
    // Iterates over chars (not bytes) for Unicode correctness, tracking
    // the UTF-8 byte offset via char.len_utf8().
    let mut token_ranges: Vec<(usize, usize)> = Vec::new();
    let mut offset: usize = 0;
    let mut run_start: Option<usize> = None;

    for ch in text.chars() {
        // Same scalar predicate as Swift: alphabetic OR ASCII digit.
        let in_token = ch.is_alphabetic() || ch.is_ascii_digit();
        if in_token {
            if run_start.is_none() {
                run_start = Some(offset);
            }
        } else if let Some(start) = run_start.take() {
            token_ranges.push((start, offset));
        }
        offset += ch.len_utf8();
    }
    if let Some(start) = run_start {
        token_ranges.push((start, offset));
    }

    if token_ranges.is_empty() {
        return Vec::new();
    }

    // Build sliding token windows.
    let mut out: Vec<(usize, usize)> = Vec::new();
    let mut index: usize = 0;
    while index < token_ranges.len() {
        let window_end = (index + window).min(token_ranges.len());
        let (first_start, _) = token_ranges[index];
        let (_, last_end) = token_ranges[window_end - 1];
        out.push((first_start, last_end - first_start));
        if window_end == token_ranges.len() {
            break;
        }
        index += stride;
    }
    out
}

// ── Cosine similarity (inline, no BLAS dependency) ───────────────────────────

/// L2-normalised cosine similarity between two equal-length f32 vectors.
///
/// Returns a value in [−1, 1]: 1.0 = identical direction, −1.0 = opposite,
/// 0.0 = orthogonal. Returns 0.0 when either vector is all-zero (no signal),
/// or when the lengths differ.
///
/// # Cross-port note
///
/// IEEE-754 f32 `+` and `*` are the same in Swift and Rust on the same
/// hardware. The result is reproducible-within-config, not universally
/// bit-identical across optimization levels.
pub fn cosine_similarity(a: &[f32], b: &[f32]) -> f32 {
    if a.len() != b.len() || a.is_empty() {
        return 0.0;
    }
    let mut dot: f32 = 0.0;
    let mut norm_a: f32 = 0.0;
    let mut norm_b: f32 = 0.0;
    for (&ai, &bi) in a.iter().zip(b.iter()) {
        dot += ai * bi;
        norm_a += ai * ai;
        norm_b += bi * bi;
    }
    let denom = norm_a.sqrt() * norm_b.sqrt();
    if denom <= 0.0 {
        0.0
    } else {
        dot / denom
    }
}
